---
title: "Figures for Github"
author: "Jesse Garcia"
date: "3/31/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```



## Figure 1
```{r cars}
library(tidyverse)
set.seed(1)

## Color blind colors
colors=c("0"="#0072B2", "-1e-04"="#F0E442", "-0.001"="#009E73", "-0.01"= "#56B4E9", "-0.1"="#E69F00", "DFE"= "#999999")

selection_levels=c("0", "-1e-04", "-0.001", "-0.01", "-0.1", "DFE")
selection_levels=levels=rev(selection_levels)

recombination_levels=c("r=1e-06", "r=1e-07", "r=1e-08" ,"r=1e-09")
dominance_levels=c("h=0.5",  "h=0")

doubletons_ld<-as_tibble(fread("/Users/jessegarcia/Documents/SLiM_ParallelRProject copy/data2/constant_population_constant_selection_constant_dfe_doubletons_all_replicates_ld.csv", data.table = FALSE)) %>%
  mutate(d=case_when(
    r_2 == 4e-04 ~ -0.0004 , 
    r_2 == 0.2399 ~ 0.0096, 
    r_2 == 1 ~ 0.0196
  ))

## Set frequency to doubletons
doubletons_ld<-doubletons_ld %>% mutate(pA=.02, pB = .02)

## Function to compute signed D'
compute_d_prime<-function(ld_df){

  ld_df<-ld_df %>% mutate(d_prime= case_when(
  d < 0 ~ d / pmin(pA*pB, (1-pA) *(1-pB)),
  d > 0 ~ d / pmin(pA*(1-pB), (1-pA) * pB  ),
  d == 0 ~ 0))
  return(ld_df)
  
}


doubletons_ld<-compute_d_prime(ld_df = doubletons_ld)
subsampled_doubletons<-doubletons_ld %>%
  filter(selection_coefficient != -0.1)  %>%
  group_by(recombination_rate, selection_coefficient, dominance_coefficient, seed) %>%
  nest() %>%
  ungroup() %>%
  group_by(recombination_rate, selection_coefficient, dominance_coefficient) %>%
  sample_n(150) %>%
  mutate(group=rep(c(1,2,3), each=50)) %>%
  unnest(data) %>% 
  ungroup()



subsampled_doubletons_summary<-subsampled_doubletons %>%
  group_by(recombination_rate, selection_coefficient, dominance_coefficient,group) %>%
  summarise(mean_d_prime=mean(d_prime))

subsampled_doubletons_summary<-subsampled_doubletons_summary %>%
  group_by(recombination_rate, selection_coefficient, dominance_coefficient) %>%
summarise(max_d_prime=max(mean_d_prime), min_d_prime=min(mean_d_prime), middle_d=sort(mean_d_prime, decreasing = FALSE)[2] , mean_of_mean_d_prime=mean(mean_d_prime)) %>% 
  ungroup() %>%
  mutate(selection_coefficient = replace(selection_coefficient, selection_coefficient == -999, "DFE")) %>%
  mutate(selection_coefficient = factor(selection_coefficient, levels=levels) ) %>%
  mutate(dominance_coefficient = glue("h={dominance_coefficient}")) %>%
  mutate(recombination_rate=glue("r={recombination_rate}")) %>%
  mutate(dominance_coefficient=factor(dominance_coefficient,levels = dominance_levels)) %>%
  mutate(recombination_rate=factor(recombination_rate, levels =recombination_levels)) 

subsampled_doubletons_summary <-subsampled_doubletons_summary %>%
  mutate(recombination_rate = fct_rev(recombination_rate))

dprime_mean_plot <- ggplot(data = subsampled_doubletons_summary, aes(x=recombination_rate, y=middle_d, colour=selection_coefficient)) +
  geom_point() +
    geom_errorbar(aes(x=recombination_rate,ymin = min_d_prime, ymax=max_d_prime) , size=3 , width=.2,alpha=0.5)  +
  facet_wrap(~dominance_coefficient , nrow=2)  +
  scale_colour_manual(values=colors) + 
  guides(fill=FALSE)  +
  theme_bw() +
    theme(strip.background =element_rect(fill="white"), text = element_text(size = 22)) +
  labs(x="Recombination rate",y=bquote("Mean D'" ), colour="Selection Coefficient") +
  geom_line(aes(group=interaction(selection_coefficient)), size=2, alpha=1)

dprime_mean_plot

ggsave(filename="../figures/fig_1_mean_dprime.tiff", plot=dprime_mean_plot, width=20, height=12)

```


## Fig 2
```{r}

library(tidyverse)
library(cowplot)

annotate_genotypes<-function(long_genotype_config, type){
  
  
  if (type == "unphased"){
  long_genotype_config<-long_genotype_config %>%
  mutate(words=case_when(
    genotype == "0/0,0/0" ~ "Homozygous reference (0/0,0/0)", 
    genotype == "0/0,0/1" ~ "Single heterozygote (0/1,0/0 or 0/0,0/1)", 
    genotype == "0/1,0/0" ~ "Single heterozygote (0/1,0/0 or 0/0,0/1)",
    genotype == "0/1,0/1" ~ "Double Heterozygote (0/1,0/1)"
  ))
  
  return(long_genotype_config)
  }
  
    
  if (type == "phased"){
  long_genotype_config<-long_genotype_config %>%
  mutate(words=case_when(
    genotype == "0|0,0|0" ~ "Homozygous reference (0|0,0|0)", 
    genotype == "0|0,0|1" ~ "Single heterozygote (0|0,0|1 & 1|0,0|0 & 0|1,0|0 & 0|0,1|0)",
    genotype == "1|0,0|0" ~ "Single heterozygote (0|0,0|1 & 1|0,0|0 & 0|1,0|0 & 0|0,1|0)",
    genotype == "0|1,0|0" ~ "Single heterozygote (0|0,0|1 & 1|0,0|0 & 0|1,0|0 & 0|0,1|0)",
    genotype == "0|0,1|0" ~ "Single heterozygote (0|0,0|1 & 1|0,0|0 & 0|1,0|0 & 0|0,1|0)",
    genotype == "0|1,0|1" ~ "Double heterozygote (0|1,0|1 & 0|1,1|0 & 1|0,1|0 & 1|0,0|1)",
    genotype == "0|1,1|0" ~ "Double heterozygote (0|1,0|1 & 0|1,1|0 & 1|0,1|0 & 1|0,0|1)",
    genotype == "1|0,1|0" ~ "Double heterozygote (0|1,0|1 & 0|1,1|0 & 1|0,1|0 & 1|0,0|1)",
    genotype == "1|0,0|1" ~ "Double heterozygote (0|1,0|1 & 0|1,1|0 & 1|0,1|0 & 1|0,0|1)"
    
  ))
  
  return(long_genotype_config)
  }
  
  
    if (type == "phased_coupling"){
  long_genotype_config<-long_genotype_config %>%
  mutate(words=case_when(
    genotype == "0|0,0|0" ~ "Homozygous reference (0|0,0|0)", 
    genotype == "0|0,0|1" ~ "Single heterozygote (0|0,0|1 & 1|0,0|0 & 0|1,0|0 & 0|0,1|0)",
    genotype == "1|0,0|0" ~ "Single heterozygote (0|0,0|1 & 1|0,0|0 & 0|1,0|0 & 0|0,1|0)",
    genotype == "0|1,0|0" ~ "Single heterozygote (0|0,0|1 & 1|0,0|0 & 0|1,0|0 & 0|0,1|0)",
    genotype == "0|0,1|0" ~ "Single heterozygote (0|0,0|1 & 1|0,0|0 & 0|1,0|0 & 0|0,1|0)",
    genotype == "0|1,0|1" ~ "Double heterozygote Coupling (0|1,0|1 & 1|0,1|0)",
    genotype == "0|1,1|0" ~ "Double heterozygote Repulsion (0|1,1|0 & 1|0,0|1)",
    genotype == "1|0,1|0" ~ "Double heterozygote Coupling (0|1,0|1 & 1|0,1|0)",
    genotype == "1|0,0|1" ~ "Double heterozygote Repulsion (0|1,1|0 & 1|0,0|1)"
    
  ))
  
  return(long_genotype_config)
  }
  
  
  
  
}




plot_rh_unphased<-function(annotated_genotypes, window_size, plot_type, allele_count_to_analyze){
  unique_words<-unique(annotated_genotypes$words) 

unique_words<-unique_words[!is.na(unique_words)]


testthat::expect_true(unique_words[3] == "Double Heterozygote (0/1,0/1)" | unique_words[3] == "Double heterozygote (0|1,0|1 & 0|1,1|0 & 1|0,1|0 & 1|0,0|1)")
if (plot_type == "bar_plot"){
doubleton_plots<-unique_words %>% map(~ { 
annotated_genotypes %>%
  filter(allele_count == allele_count_to_analyze, !is.na(words), words ==.x)  %>%
      mutate(distance_breaks=cut_interval(distance, length=window_size)) %>%
  group_by(distance_breaks, variation_type, recombination_rate, words) %>%
  summarise(mean_count=mean(count)) %>% 
    ungroup() %>%
  ggplot(aes(x=distance_breaks, y=mean_count, fill=variation_type)) +
  geom_col(position="dodge2") +
  facet_grid(~words,scales="free") + 
    theme_bw()
} ) } else if (plot_type == "line_plot") {
  
  doubleton_plots<-unique_words %>% map(~ { 
annotated_genotypes %>%
  filter(allele_count == allele_count_to_analyze, !is.na(words), words ==.x)  %>%
      mutate(distance_breaks=cut_interval(distance, length=window_size)) %>%
  group_by(distance_breaks, variation_type, recombination_rate, words) %>%
  summarise(mean_count=mean(count)) %>% 
    ungroup() %>%
  ggplot(aes(x=distance_breaks, y=mean_count, colour=variation_type, group=variation_type)) +
  ylab(bquote('Mean '*H[R]^.(allele_count_to_analyze))) +
  labs( x="Physical distance (kbp)", colour="Variation") +
  geom_point(size=2) +
  geom_line(size=1.5) +
#  facet_grid(~words,scales="free") + 
  theme_bw() +
  theme(text=element_text(size=16)) +
  scale_color_manual(values=c("Synonymous"="dodgerblue1", "Nonsynonymous"="darkorchid2"))
}  )
}


  
return(doubleton_plots)
}



## Recombination of 1e-09
# path on computer "/Users/jessegarcia/Documents/SLiM_ParallelRProject copy/data2/configuration_gravel_simulations_low_recomb_ac_1_2.rds"
simulation_df <- read_rds("configuration_gravel_simulations_low_recomb_ac_1_2.rds") %>% mutate_if(is.numeric, replace_na,  0) %>%
  mutate(distance=abs(pos_1-pos_2)) %>%
  mutate(variation_type=case_when(
    variation_type == 1 ~ "Nonsynonymous",
    variation_type == 2 ~ "Synonymous"
  )) %>%
  mutate(variation_type=as.factor(variation_type)) %>% filter(recombination_rate==1e-09 )

long_simulation_df_genotype_config <- gather(simulation_df, 
                                 "0|0,0|0", "0|0,0|1", "1|0,0|0", "0|0,1|0", "0|1,0|0", "1|0,0|1", "1|0,1|0", "0|1,1|0", "0|1,0|1", "1|1,0|0" ,"0|0,1|1" ,"1|1,1|1" , 
                                 value="count", 
                                 key="genotype") 


long_simulation_df_genotype_config <- annotate_genotypes(long_simulation_df_genotype_config, type="phased")


annotated_genotypes_recomb_1e9 <- long_simulation_df_genotype_config


## Recombination of 1e-08
# path on computer "/Users/jessegarcia/Desktop/SLiM_ParallelRProject/data2/configuration_gravel_simulations_average_recomb_ac_1_2.rds"
simulation_df <- read_rds("configuration_gravel_simulations_average_recomb_ac_1_2.rds") %>% mutate_if(is.numeric, replace_na,  0) %>%
  mutate(distance=abs(pos_1-pos_2)) %>%
  mutate(variation_type=case_when(
    variation_type == 1 ~ "Nonsynonymous",
    variation_type == 2 ~ "Synonymous"
  )) %>%
  mutate(variation_type=as.factor(variation_type)) %>% filter(recombination_rate==1e-08 )

long_simulation_df_genotype_config <- gather(simulation_df, 
                                 "0|0,0|0", "0|0,0|1", "1|0,0|0", "0|0,1|0", "0|1,0|0", "1|0,0|1", "1|0,1|0", "0|1,1|0", "0|1,0|1", "1|1,0|0" ,"0|0,1|1" ,"1|1,1|1" , 
                                 value="count", 
                                 key="genotype") 


long_simulation_df_genotype_config <- annotate_genotypes(long_simulation_df_genotype_config, type="phased")


annotated_genotypes_recomb_1e8 <- long_simulation_df_genotype_config





## Simulations

#Recombination 1e-09
simulations_doubletons_hr_recomb_1e9 <- plot_rh_unphased(annotated_genotypes_recomb_1e9, window_size = 1500, plot_type = "line_plot" , allele_count_to_analyze = 2)[[3]]
simulations_singletons_hr_recomb_1e9 <- plot_rh_unphased(annotated_genotypes_recomb_1e9, window_size = 1500, plot_type = "line_plot" , allele_count_to_analyze = 1)[[3]]
#Recombination 1e-08
simulations_doubletons_hr_recomb_1e8 <- plot_rh_unphased(annotated_genotypes_recomb_1e8, window_size = 1500, plot_type = "line_plot" , allele_count_to_analyze = 2)[[3]]
simulations_singletons_hr_recomb_1e8 <- plot_rh_unphased(annotated_genotypes_recomb_1e8, window_size = 1500, plot_type = "line_plot" , allele_count_to_analyze = 1)[[3]]



# extract a legend that is laid out horizontally
legend_b <- get_legend(
  simulations_doubletons_hr_recomb_1e8 + 
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom")
)




hr_plots <- plot_grid(
simulations_singletons_hr_recomb_1e9 + theme(legend.position = "none") + scale_x_discrete(labels=c("1.5", "", "" ,"6" ,"","","10"))  + xlab(NULL)+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + labs(title="r=1e-09"), 
simulations_singletons_hr_recomb_1e8 + theme(legend.position = "none") + scale_x_discrete(labels=c("1.5", "", "" ,"6" ,"","","10")) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + ylab(" ")  + labs(title="r=1e-08"),
simulations_doubletons_hr_recomb_1e9 + theme(legend.position = "none") + scale_x_discrete(labels=c("1.5", "", "" ,"6" ,"","","10")) + xlab("Physical distance (kbp)")  ,
simulations_doubletons_hr_recomb_1e8 + theme(legend.position = "none") + scale_x_discrete(labels=c("1.5", "", "" ,"6" ,"","","10"))+ xlab("Physical distance (kbp)")  + ylab(" "),
labels = c('A', 'C', 'B', 'D'), label_size = 12, nrow=2, ncol=2)



simulated_hr_plots<-plot_grid(hr_plots,legend_b, ncol=1, rel_heights = c(1, .1))
simulated_hr_plots
ggsave(filename = "../figures/figure_2_simulated_hr_plots.tiff", width=20, height=12)
```


## Fig3a
```{r}
library(tidyverse)
library(data.table)
library(glue)
library(mltools)

se <- function(x) sqrt(var(x)/length(x))


yri_df_b<-as_tibble(read_rds(file ="/Users/jessegarcia/Documents/SLiM_ParallelRProject/data2/YRI_b_value_breaks_5_physical_breaks_200.rds")) %>%
  dplyr::rename(D=X6,variation=variation_type, r_square=genetic_distance, genetic_distance=X3, DPrime=X7) %>%
  mutate(Variation=glue("={variation}"), rSquare=r_square) %>% mutate(population="YRI") %>%  mutate(Variation=case_when(
  Variation == "=nonsynonymous_SNV" ~ "Nonsynonymous", 
  Variation == "=synonymous_SNV" ~ "Synonymous" 
)) 

yri_df_b$GeneticDistanceBreaks<-bin_data(yri_df_b$genetic_distance, bins=10, binType="quantile")

yri_plot<-yri_df_b %>%
    group_by(GeneticDistanceBreaks, Variation) %>%
  summarise(mean_d=mean(DPrime), 
            sd_d=se(DPrime)) %>%
  ggplot(aes(x=GeneticDistanceBreaks, y=mean_d, colour=Variation, group=Variation)) +
  geom_errorbar(aes(ymin=mean_d-sd_d, ymax=mean_d+sd_d), width=.2) +
  geom_point(size=1.5) + 
  geom_line(size=1) +
  theme_bw() +
  labs(y="Mean D' of Matched Pairs", x="Genetic Distance Bins (cM)") + 
  scale_color_manual(values=c("Synonymous"="dodgerblue1", "Nonsynonymous"="darkorchid2")) +
  scale_x_discrete(labels=c("1st quantile", "2nd quantile", "3rd quantile", "4th quantile", "5th quantile", "6th quantile"))+
  theme(text=element_text(size=22))





mean_dprime_of_matched_pairs <- yri_plot + theme(legend.position = "bottom")

mean_dprime_of_matched_pairs
ggsave(filename = "../figures/figure_3a_mean_dprime_of_matched_pairs.tiff", width=20, height=12)

```
## Fig 3b

```{r}
library(tidyverse)
library(data.table)
library(glue)

prepMatchedDataForPermutation <- function(MatchedLDData){
  MatchedLDData$Variation <- as.character(MatchedLDData$Variation)
  MatchedLDData$Variation[MatchedLDData$Variation == "=nonsynonymous_SNV"] <- "Nonsynonymous SNV"
  MatchedLDData$Variation[MatchedLDData$Variation == "=synonymous_SNV"] <- "Synonymous SNV"
  return(MatchedLDData)
}

permuteLDStat <- function(preppedMatchedData, SummaryStat){
  nonsynonymousStats <- preppedMatchedData[[SummaryStat]][preppedMatchedData$Variation == "Nonsynonymous SNV" ]
  synonymousStats <- preppedMatchedData[[SummaryStat]][preppedMatchedData$Variation == "Synonymous SNV" ]
  permute(nonsynonymousStats=nonsynonymousStats, synonymousStats=synonymousStats, statistic = SummaryStat )
  
}

plottingDPermutations <- function(MatchedLDData, LDStat="D"){
  preppedMatchedData <- prepMatchedDataForPermutation(MatchedLDData)
  permuteLDStat(preppedMatchedData=preppedMatchedData,SummaryStat= LDStat) + theme(legend.position="none")
}


oneTest <- function(nonsynonymousStats, synonymousStats){
  differences <- nonsynonymousStats - synonymousStats
  #For Randomization test on matched samples you need to randomly change the sign of the differences. 
  #Sample 1 and -1, multiply by differences
  signs <- sample(size=length(differences), c(1,-1), replace = TRUE)
  newsample <- signs*differences
  randomizedMean <- mean(newsample)
  return(randomizedMean)
}

permute<-function(nonsynonymousStats,synonymousStats, replications = 10000, statistic="Some LD Statistic" ){
  actualDiff<-mean(nonsynonymousStats - synonymousStats)
  permutations<-replicate(replications, oneTest(nonsynonymousStats, synonymousStats))
  if (statistic == "rSquare"){
    title <- bquote( ~ R^2)
  } else if (statistic == "DPrime") {
    title <- bquote( ~ "D' ")
  } else if (statistic == "D"){
    title <- bquote(~ "D ")
  } else if (statistic == "FreqAB"){
    title <- bquote(~ "Count of AB Haplotypes")
  }
  permDF<- as.data.frame(permutations)
  names(permDF)<-"MeanDif"
  ggplot(permDF, aes(x=MeanDif)) + 
    geom_histogram() + 
    labs(title=title, x="Mean Difference \n Between Matched Pairs", y="Frequency") + 
    geom_vline(aes(xintercept = actualDiff, color = "Observed Mean Difference"), linetype=2, size=2) + 
    theme_bw()
}


prepMatchedDataForPermutation<-function(MatchedLDData){
  MatchedLDData$Variation <- as.character(MatchedLDData$Variation)
  MatchedLDData$Variation[MatchedLDData$Variation == "=nonsynonymous_SNV"] <- "Nonsynonymous SNV"
  MatchedLDData$Variation[MatchedLDData$Variation == "=synonymous_SNV"] <- "Synonymous SNV"
  return(MatchedLDData)
}

permuteLDStat<-function(preppedMatchedData, SummaryStat){
  nonsynonymousStats <- preppedMatchedData[[SummaryStat]][preppedMatchedData$Variation == "Nonsynonymous SNV" ]
  synonymousStats <- preppedMatchedData[[SummaryStat]][preppedMatchedData$Variation == "Synonymous SNV" ]
  permute(nonsynonymousStats=nonsynonymousStats, synonymousStats=synonymousStats, statistic = SummaryStat )
  
}

plottingDPermutations <- function(MatchedLDData, LDStat="D"){
  preppedMatchedData <- prepMatchedDataForPermutation(MatchedLDData)
  permuteLDStat(preppedMatchedData=preppedMatchedData,SummaryStat= LDStat) + theme(legend.position="none")
}

plottingDPrimePermutations <- function(MatchedLDData, LDStat="DPrime"){
  preppedMatchedData <- prepMatchedDataForPermutation(MatchedLDData)
  permuteLDStat(preppedMatchedData=preppedMatchedData,SummaryStat= LDStat) + labs(color="") + theme(legend.position = c(0.89,.8)) + theme(legend.position="none")
  
}

plottingRSquarePermutations <- function(MatchedLDData, LDStat="rSquare"){
  preppedMatchedData <- prepMatchedDataForPermutation(MatchedLDData)
  permuteLDStat(preppedMatchedData=preppedMatchedData,SummaryStat= LDStat) + theme(legend.position="none")
}

plottingFreqABPermutations <- function(MatchedLDData, LDStat="FreqAB"){
  preppedMatchedData <- prepMatchedDataForPermutation(MatchedLDData)
  permuteLDStat(preppedMatchedData=preppedMatchedData,SummaryStat= LDStat) + theme(legend.position="none")
}



plotPermutations3<-function(MatchedLDData, population){
  DPermGraph <- plottingDPermutations(MatchedLDData = MatchedLDData) +theme(text = element_text(size=14)) +   scale_x_continuous(labels = scales::scientific, breaks = scales::pretty_breaks(n=3)) + geom_histogram( fill="slategray3") + xlab(NULL)
  DPrimePermGraph <- plottingDPrimePermutations(MatchedLDData = MatchedLDData) +theme(text = element_text(size=14))+   scale_x_continuous(labels = scales::scientific, breaks = scales::pretty_breaks(n=3))  + geom_histogram( fill="slategray3") + xlab(NULL)
  RSquarePermGraph <- plottingRSquarePermutations(MatchedLDData = MatchedLDData)+theme(text = element_text(size=14))+   scale_x_continuous(labels = scales::scientific, breaks = scales::pretty_breaks(n=3)) + geom_histogram( fill="slategray3") + xlab("Mean Difference")
  p<-plot_grid(DPermGraph  +
                 theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                       panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")), 
               DPrimePermGraph +
                 theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                       panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")), 
               RSquarePermGraph +
                 theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                       panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")), ncol=1)
  p
}




theme_set((theme_bw(base_size=22)))



set.seed(1)

## Path on computer /Users/jessegarcia/Documents/garcia_and_lohmueller_2020/fig_4_updated/YRI_b_value_breaks_5_physical_breaks_200.rds
yri_df_b<-as_tibble(read_rds("YRI_b_value_breaks_5_physical_breaks_200.rds")) %>%
  dplyr::rename(D=X6,variation=variation_type, r_square=genetic_distance, genetic_distance=X3, DPrime=X7) %>%
  mutate(Variation=glue("={variation}"), rSquare=r_square) %>% mutate(population="YRI") 




permutation_test_figure <- plotPermutations3(yri_df_b, population="YRI")

permutation_test_figure

ggsave("../figures/figure_3b_mean_dprime_of_matched_pairs.tiff", width=20, height=12)
```


## Fig 4

```{r}
library(tidyverse)
library(janitor)
library(data.table)
library(testthat)
library(glue)
library(cowplot)
library(binr)
library(mltools)
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")


ceu_df_b <- as_tibble(read_rds(file = "ceu_b_value_breaks_5_physical_breaks_200.rds")) %>%
  dplyr::rename(D=X6,variation=variation_type, r_square=genetic_distance, genetic_distance=X3, DPrime=X7) %>%
  mutate(Variation=glue("={variation}"), rSquare=r_square) %>% mutate(population="CEU")
chb_df_b <- as_tibble(read_rds(file ="chb_b_value_breaks_5_physical_breaks_200.rds")) %>%
  dplyr::rename(D=X6,variation=variation_type, r_square=genetic_distance, genetic_distance=X3, DPrime=X7) %>%
  mutate(Variation=glue("={variation}"), rSquare=r_square) %>% mutate(population="CHB")
mxl_df_b <- as_tibble(read_rds(file = "MXL_b_value_breaks_5_physical_breaks_200.rds")) %>%
  dplyr::rename(D=X6,variation=variation_type, r_square=genetic_distance, genetic_distance=X3, DPrime=X7) %>%
  mutate(Variation=glue("={variation}"), rSquare=r_square) %>% mutate(population="MXL")
jpt_df_b <- as_tibble(read_rds(file ="JPT_b_value_breaks_5_physical_breaks_200.rds")) %>%
  dplyr::rename(D=X6,variation=variation_type, r_square=genetic_distance, genetic_distance=X3, DPrime=X7) %>%
  mutate(Variation=glue("={variation}"), rSquare=r_square) %>% mutate(population="JPT")
yri_df_b <- as_tibble(read_rds(file ="YRI_b_value_breaks_5_physical_breaks_200.rds")) %>%
  dplyr::rename(D=X6,variation=variation_type, r_square=genetic_distance, genetic_distance=X3, DPrime=X7) %>%
  mutate(Variation=glue("={variation}"), rSquare=r_square) %>% mutate(population="YRI")


#yri_df_b
empirical_matched <- bind_rows(ceu_df_b, chb_df_b,mxl_df_b , jpt_df_b)


plot_ld_across_genetic_distance <- function(ACFilteredLD,AlleleCountMax, bins, removeZeroCM){
  
  if(removeZeroCM){
    doubletonsLD<- na.omit(ACFilteredLD %>% filter( AC <= AlleleCountMax & GeneticDistance != 0))
  } else {
  doubletonsLD<- na.omit(ACFilteredLD %>% filter( AC <= AlleleCountMax))
}

doubletonsLD$GeneticDistanceBreaks <- bin_data(doubletonsLD$GeneticDistance, bins=bins, binType="quantile")


percentileNames <- c("0-10%", "10-20%", "20-30%", "30-40%", "40-50%", "50-60%", "60-70%", "70-80%", "80-90%", "90-100%")


doubletonsLD <- doubletonsLD %>% 
  mutate(Variation=case_when(
    Variation=="=nonsynonymous_SNV" ~ "Nonsynonymous",
    Variation=="=synonymous_SNV" ~ "Synonymous"
  ))

return(doubletonsLD)

}

  
empirical_matched$GeneticDistanceBreaks <- bin_data(empirical_matched$genetic_distance, bins=10, binType="quantile")

empirical_matched <- empirical_matched %>%  
  mutate(Variation=case_when(
  Variation == "=nonsynonymous_SNV" ~ "Nonsynonymous", 
  Variation == "=synonymous_SNV" ~ "Synonymous" 
))


figure_4_a <- empirical_matched %>% 
  group_by(GeneticDistanceBreaks, Variation, population) %>%
  summarise(mean_d=mean(DPrime)) %>%
  ggplot(aes(x=GeneticDistanceBreaks, y=mean_d, colour=Variation, group=Variation)) +
  geom_point(size=1.5) + 
  geom_line(size=1) +
  theme_bw() +
  facet_wrap(~population, ncol=1) + 
  labs(y="Mean D' of Matched Pairs", x="Genetic Distance Bins (cM)") + 
  scale_color_manual(values=c("Synonymous"="dodgerblue1", "Nonsynonymous"="darkorchid2")) +
  scale_x_discrete(labels=c("3.47e-05", "3.32e-04", "1.56e-03", "6.46e-03", "6.28e-01")) +
  theme(text=element_text(size=18))

figure_4_a

ggsave("../figures/figure_4a_mean_dprime_of_matched_pairs_across_populations.tiff", width=20, height=12)


```

## Fig 5

```{r}
library(tidyverse)
library(cowplot)


high_coverage_hg38_genotype_configuration <- read_csv("/Users/jessegarcia/Documents/SLiM_ParallelRProject copy/data2/high_coverage_hg38_genotype_configuration.csv")

long_low_genotype_config <- read_csv("/Users/jessegarcia/Desktop/SLiM_ParallelRProject/data2/long_low_genotype_config.csv")

plot_rh_unphased<-function(annotated_genotypes, window_size, plot_type, allele_count_to_analyze){
  unique_words<-unique(annotated_genotypes$words) 
  
  unique_words<-unique_words[!is.na(unique_words)]
  
  
  testthat::expect_true(unique_words[3] == "Double Heterozygote (0/1,0/1)" | unique_words[3] == "Double heterozygote (0|1,0|1 & 0|1,1|0 & 1|0,1|0 & 1|0,0|1)")
  if (plot_type == "bar_plot"){
    doubleton_plots<-unique_words %>% map(~ { 
      annotated_genotypes %>%
        filter(allele_count == allele_count_to_analyze, !is.na(words), words ==.x)  %>%
        mutate(distance_breaks=cut_interval(distance, length=window_size)) %>%
        group_by(distance_breaks, variation_type, recombination_rate, words) %>%
        summarise(mean_count=mean(count)) %>% 
        ungroup() %>%
        ggplot(aes(x=distance_breaks, y=mean_count, fill=variation_type)) +
        geom_col(position="dodge2") +
        facet_grid(~words,scales="free") + 
        theme_bw()
    } ) } else if (plot_type == "line_plot") {
      
      doubleton_plots<-unique_words %>% map(~ { 
        annotated_genotypes %>%
          filter(allele_count == allele_count_to_analyze, !is.na(words), words ==.x)  %>%
          mutate(distance_breaks=cut_interval(distance, length=window_size)) %>%
          group_by(distance_breaks, variation_type, recombination_rate, words) %>%
          summarise(mean_count=mean(count)) %>% 
          ungroup() %>%
          ggplot(aes(x=distance_breaks, y=mean_count, colour=variation_type, group=variation_type)) +
          ylab(bquote('Mean '*H[R]^.(allele_count_to_analyze))) +
          labs( x="Physical distance (kbp)", colour="Variation") +
          geom_point(size=2) +
          geom_line(size=1.5) +
          #  facet_grid(~words,scales="free") + 
          theme_bw() +
          theme(text=element_text(size=16)) +
          scale_color_manual(values=c("Synonymous"="dodgerblue1", "Nonsynonymous"="darkorchid2"))
      }  )
    }
  
  
  
  return(doubleton_plots)
}

## High coverage hg38

highcoverage_hg38_doubletons_hr<-plot_rh_unphased(high_coverage_hg38_genotype_configuration %>% mutate(recombination_rate="unknown")%>% 
  mutate(variation_type=case_when(
  variation_type == "nonsynonymous_SNV" ~ "Nonsynonymous", 
  variation_type == "synonymous_SNV" ~ "Synonymous" 
)) , window_size = 1500, plot_type = "line_plot" , allele_count_to_analyze = 2)[[3]]




## low coverage hg19
low_coverage_doubletons_hr<-plot_rh_unphased(long_low_genotype_config %>% mutate(recombination_rate="unknown")%>% 
  mutate(variation_type=case_when(
  variation_type == "nonsynonymous_SNV" ~ "Nonsynonymous", 
  variation_type == "synonymous_SNV" ~ "Synonymous" 
)) , window_size = 1500, plot_type = "line_plot" , allele_count_to_analyze = 2)[[3]]




# extract a legend that is laid out horizontally
legend_b <- get_legend(
  low_coverage_doubletons_hr + 
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom")
)




### just empirical

## High coverage hg38

highcoverage_hg38_singletons_hr<-plot_rh_unphased(high_coverage_hg38_genotype_configuration %>% mutate(recombination_rate="unknown")%>% 
  mutate(variation_type=case_when(
  variation_type == "nonsynonymous_SNV" ~ "Nonsynonymous", 
  variation_type == "synonymous_SNV" ~ "Synonymous" 
)) , window_size = 1500, plot_type = "line_plot" , allele_count_to_analyze = 1)[[3]]




## low coverage hg19
low_coverage_singletons_hr<-plot_rh_unphased(long_low_genotype_config %>% mutate(recombination_rate="unknown")%>% 
  mutate(variation_type=case_when(
  variation_type == "nonsynonymous_SNV" ~ "Nonsynonymous", 
  variation_type == "synonymous_SNV" ~ "Synonymous" 
)) , window_size = 1500, plot_type = "line_plot" , allele_count_to_analyze = 1)[[3]]



hr_plots<-plot_grid(
low_coverage_doubletons_hr + theme(legend.position = "none") + xlab(NULL)+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + labs(title="Low coverage 1KGP"),
highcoverage_hg38_doubletons_hr + theme(legend.position = "none")+ scale_x_discrete(labels=c("1.5", "", "" ,"6" ,"","","10"))+ theme(axis.text.x = element_text(size=14, angle =45, hjust=1))  +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + ylab(" ")  + labs(title="High coverage NYGC 1KGP"), 
low_coverage_singletons_hr+ scale_x_discrete(labels=c("1.5", "", "" ,"6" ,"","","10"))  + theme(legend.position = "none"),
highcoverage_hg38_singletons_hr+ scale_x_discrete(labels=c("1.5", "", "" ,"6" ,"","","10"))  + theme(legend.position = "none") + ylab(" "),

labels = c('A', 'C', 'B', 'D'), label_size = 12, nrow=2, ncol=2)



plot_grid(hr_plots,legend_b, ncol=1, rel_heights = c(1, .1))

ggsave("../figures/figure_5_hr_plots.tiff", width=20, height=12)

```



## Fig 6

```{r}
library(data.table)
library(tidyverse)
library(glue)
library(binr)
library(mltools)
library(janitor)

## Colorblind palette
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")



midpoints <- function(x, dp=2){
lower <- as.numeric(gsub(",.*","",gsub("\\(|\\[|\\)|\\]","", x)))
upper <- as.numeric(gsub(".*,","",gsub("\\(|\\[|\\)|\\]","", x)))
return(round(lower+(upper-lower)/2, dp))
}



getCutBreaks <- function(x, dp=30){
lower <- as.numeric(gsub(",.*","",gsub("\\(|\\[|\\)|\\]","", x)))
upper <- tail(as.numeric(gsub(".*,","",gsub("\\(|\\[|\\)|\\]","", x))),n=1)
return(c(lower, upper))
}


getBins <- function(x, dp=30){
lower <- as.numeric(gsub(",.*","",gsub("\\(|\\[|\\)|\\]","", x)))
upper <- as.numeric(gsub(".*,","",gsub("\\(|\\[|\\)|\\]","", x)))




bins<-tibble(
  lower=lower,
  upper=upper
)

return(bins)
}


findTheNumberOfVariantsPerAC <- function(empiricalDf) {
  
  counts <- empiricalDf %>% 
    group_by(ac, geneticDistanceBreaks, variation) %>%
    count()
  
  return(counts)
  
}


binSimulatedIntoEmpiricalBreaks <- function(binnedEmpiricalBreaks, simulatedGeneticDistance) {
  bins <- getBins(levels(binnedEmpiricalBreaks))
  simulatedGeneticDistanceBreaks <- bin_data(simulatedGeneticDistance, boundaryType = "lcro]", bins=bins) 

  return(simulatedGeneticDistanceBreaks)
  
  
}


sampleFromCounts <- function(binnedSimulation, empiricalCounts){
  samples <- list()
  for (i in 1:nrow(empiricalCounts) ){
    
    AC <- empiricalCounts$ac[i]
    geneticDistanceBreaks <- empiricalCounts$geneticDistanceBreaks[i]
    variation <- empiricalCounts$variation[i]
    sampleSize <- empiricalCounts$n[i]
    
    
    test <- binnedSimulation %>%
      filter(ac == AC, geneticDistanceBreaks == !!geneticDistanceBreaks, variation == !!variation) %>%
      sample_n(sampleSize)
    
    samples[[i]] <- test
  }
  
  samples <- bind_rows(samples)
  return(samples)
}


differenceInD <- function(df){
  nsD <- df %>% filter(variation == "Nonsynonymous") %>% pull(mean_d)
  sD <- df %>% filter(variation == "Synonymous" ) %>% pull(mean_d)
  
  difference <- nsD-sD
  ratio <- nsD/sD
  normalized_difference <- ( mean(nsD) - mean(sD) ) / mean(sD)
  
  ldDf <- tibble(
    difference=difference,
    ratio=ratio,
    normalized_difference=normalized_difference,
    nsD=nsD, 
    sD=sD  )
  
  return(ldDf)
  
}


binEmpiricalIntoSpecificEmpiricalBreaks <- function(binnedEmpiricalBreaks, empiricalGeneticDistance) {

   
  bins <- getBins(levels(binnedEmpiricalBreaks))
  empiricalGeneticDistanceBreaks <- bin_data(empiricalGeneticDistance, boundaryType = "lcro]", bins=bins) 

  return(empiricalGeneticDistanceBreaks)
  
  
}


mean_normalized_difference_in_d <- function(df, population){
  df %>% 
  ungroup() %>% 
  filter(genetic_distance != 0) %>%
  mutate(geneticDistanceBreaks=bin_data(genetic_distance, bins=5,binType = "quantile" )) %>%
  mutate(variation=case_when(variation=="synonymous_SNV" ~ "Synonymous", variation=="nonsynonymous_SNV" ~ "Nonsynonymous"    )) %>%
  group_by(variation, geneticDistanceBreaks) %>% add_tally() %>%
  summarise(mean_d=mean(d),count=unique(n) )  %>%  
  ungroup() %>% 
  group_by(geneticDistanceBreaks) %>%
  group_map( ~ differenceInD(.x) ) %>%
  ggplot(aes(x=geneticDistanceBreaks, y=normalized_difference, group=1)) +
  geom_line() +
  labs(x="Genetic distance quantiles (cM)", y="Mean normalized difference in D (NS - S) / S ", colour="", subtitle=population) +
  theme_bw() +
  geom_hline(yintercept = 0, colour="grey", size=2, linetype="dashed")
  
  
}

bin_one_population_into_another_compute_mean_difference <- function(ld_df,population_of_ld_df, empiricalDf) {
  
  ld_df$geneticDistanceBreaks <- binEmpiricalIntoSpecificEmpiricalBreaks( binnedEmpiricalBreaks = empiricalDf$geneticDistanceBreaks, 
                                           empiricalGeneticDistance = ld_df$genetic_distance)
  
  ld_df %>% 
    ungroup() %>% 
    filter(genetic_distance != 0) %>%
    mutate(variation=case_when(variation=="synonymous_SNV" ~ "Synonymous", variation=="nonsynonymous_SNV" ~ "Nonsynonymous"    )) %>%
    group_by(variation, geneticDistanceBreaks) %>% 
    add_tally() %>%
    summarise(mean_d=mean(d), count_of_pairs=unique(n) )  %>%  
    ungroup() %>% 
    group_by(geneticDistanceBreaks) %>%
    group_map( ~ differenceInD(.x) ) %>%
    mutate(population=population_of_ld_df, type=parse_character(glue("Empirical ({population_of_ld_df})"))) %>% 
    ungroup()

}


differenceInD_2 <- function(df){
  nsD <- df %>% filter(variation == "Nonsynonymous") %>% pull(mean_d)
  sD <- df %>% filter(variation == "Synonymous" ) %>% pull(mean_d)
  
  difference <- nsD-sD
  ratio <- nsD/sD
  normalized_difference <- ( mean(nsD) - mean(sD) ) / mean(sD)
  count_of_ns_pairs <- df %>% filter(variation == "Nonsynonymous") %>% pull(count_of_pairs)
  count_of_s_pairs <- df %>% filter(variation == "Synonymous") %>% pull(count_of_pairs)

  ldDf <- tibble(
    difference=difference,
    ratio=ratio,
    normalized_difference=normalized_difference,
    nsD=nsD, 
    sD=sD,
    count_of_ns_pairs=count_of_ns_pairs,
    count_of_s_pairs=count_of_s_pairs
  )
  
  return(ldDf)
  
}

ACFilteredLD <- readRDS("/Users/jessegarcia/Documents/New LD Stats D Prime Tables/ACFilteredLDWithGeneticDistanceJan18.RDS")
matchedPhysicalBreaks50Data <- readRDS("/Users/jessegarcia/Documents/New LD Stats D Prime Tables/matchedPhysicalBreaks50Data.rds")

mergedEmpirical <- as_tibble(merge(ACFilteredLD, matchedPhysicalBreaks50Data)) %>% 
  filter(AC <= 5, GeneticDistance != 0) %>% 
  clean_names() %>% 
  mutate(variation=case_when(
  variation == "=nonsynonymous_SNV" ~ "Nonsynonymous", 
  variation == "=synonymous_SNV" ~ "Synonymous" 
))

  

empiricalDf <- mergedEmpirical %>%
  mutate(geneticDistanceBreaks=bin_data(genetic_distance, bins=5, binType="quantile"))


empiricalCounts <- findTheNumberOfVariantsPerAC(empiricalDf = empiricalDf)
set.seed(125)

samples <- read_rds("/Users/jessegarcia/Desktop/SLiM_ParallelRProject/data/resamples_small_recomb_rates_ac_1_5_june_29_2019.rds")





test <- samples %>%
  group_by(resample, variation, geneticDistanceBreaks) %>%
  summarise(mean_d=mean(d)) %>%
  group_by(resample, geneticDistanceBreaks) %>%
  group_modify( ~ differenceInD(.x) )

test$midpoints <- midpoints(test$geneticDistanceBreaks, dp=30)


summaryOfDifference <- test %>%
  ungroup() %>%
  group_by(geneticDistanceBreaks) %>%
  summarise(median_difference=median(difference), 
            median_ratio=median(ratio),
            mean_difference=mean(difference),
            mean_ratio=mean(ratio),
            mean_normalized_difference=mean(normalized_difference)) %>%
  ungroup() 

empiricalDifferences <- empiricalDf  %>%
  group_by(variation, geneticDistanceBreaks) %>%
  summarise(mean_d=mean(d))  %>%  
  ungroup() %>% 
  group_by(geneticDistanceBreaks) %>%
  group_modify( ~ differenceInD(.x) )

neutralModel <- tibble(
  difference=0, 
  type="Neutral", 
  resample=4444, 
  alpha=1,
  geneticDistanceBreaks=summaryOfDifference$geneticDistanceBreaks,
  size=1, 
  colour="#999999",
  ratio=1,
  normalized_difference=0
)



graphDf <- bind_rows(test %>% mutate(type="Resample mean", alpha=0.4, size=1, colour="#E69F00") , 
                   summaryOfDifference %>% mutate(difference=mean_difference,ratio=mean_ratio,normalized_difference=mean_normalized_difference, type="Simulated", resample=9999, alpha=1, size=1.1, colour="#E69F00") ,
                   empiricalDifferences %>% mutate(type="Empirical (YRI)" , resample=10000, alpha=1, size=1.1,colour="#009E73" ),
                   neutralModel
                   ) 




cols <- c("Simulated" = "#E69F00", "Empirical (YRI)" = "#009E73", "Resample mean" ="#E69F00", "Neutral" ="#999999")
lines <- c("Simulated" = "solid", "Empirical (YRI)" = "solid", "Resample mean" ="solid", "Neutral" ="dashed")
alphas <- c("Simulated" = 1, "Empirical (YRI)" = 1, "Resample mean" =0.19, "Neutral" =1)
sizes<-c("Simulated" = 3, "Empirical (YRI)" = 3, "Resample mean" =1, "Neutral" =2)
xbreaks<-c("[1.62011535221e-11, 6.35875037262e-05)" = "[1.62e-11, 6.36e-05)",
           "[6.35875037262e-05, 0.000597266333334)" = "[6.36e-05, 5.97e-04)",
           "[0.000597266333334, 0.00252735883636)" ="[5.97e-04, 2.53e-03)",
            "[0.00252735883636, 0.00928393376262)" = "[2.53e-03, 9.28e-03)",
            "[0.00928393376262, 0.732906302888]" =  "[9.28e-03, 0.733]"
           )



graphDf %>%
  ggplot(aes(x=geneticDistanceBreaks, y=normalized_difference, group=resample, color=type, alpha=type, size=type, linetype=type)) +
  geom_line() +
  labs(x="Genetic distance quantiles (cM)", y="Mean normalized difference in D (NS - S) / S ", colour="") +
  scale_color_manual(values=cols, breaks=c("Empirical (YRI)", "Neutral", "Simulated")) +
  scale_linetype_manual(values=lines, guide=F) +
  scale_alpha_manual(values=alphas, guide=F) +
  scale_size_manual(values=sizes, guide=F) + 
  guides(colour = guide_legend(override.aes = list(size = 2))) +
  scale_x_discrete(labels=xbreaks) + 
  theme(legend.position = "bottom")


ggsave("../figures/figure_6_mean_difference_in_d.tiff", width=20, height=12)

```


## SI Fig 2
```{r}
library(data.table)
library(tidyverse)
library(glue)


## This data file can be found at https://datadryad.org/stash/dataset/doi:10.5068/D1G674. Unzip the file for fread to work. 
constant_ld<-as_tibble(fread("/Users/jessegarcia/Documents/SLiM_ParallelRProject/data2/constant_dfe_and_constant_selection_ld.csv", data.table = FALSE))

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")


selection_levels=c("0", "-1e-04", "-0.001", "-0.01", "-0.1", "DFE")
selection_levels=levels=rev(selection_levels)

## Changing the order of dominance levels for this
recombination_levels=c("r=1e-06", "r=1e-07", "r=1e-08" ,"r=1e-09")
dominance_levels=c("h=0.5", "h=0" )



summarized_constant_ld=constant_ld %>%
  mutate(distance_breaks=cut_interval(x=dist, length=1000, labels=FALSE)) %>%
  group_by(recombination_rate, selection_coefficient, dominance_coefficient,seed, distance_breaks) %>%
  summarise(mean_r_2=mean(r_2)) %>% 
  ungroup() %>%
  mutate(selection_coefficient = replace(selection_coefficient, selection_coefficient == -999, "DFE")) %>%
  mutate(selection_coefficient = factor(selection_coefficient, levels=levels) ) %>%
  mutate(dominance_coefficient = glue("h={dominance_coefficient}")) %>%
  mutate(recombination_rate=glue("r={recombination_rate}")) %>%
  mutate(dominance_coefficient=factor(dominance_coefficient,levels= dominance_levels)) %>%
  mutate(recombination_rate=factor(recombination_rate, levels=recombination_levels)) %>%
  group_by(recombination_rate, selection_coefficient, dominance_coefficient, distance_breaks) %>%
  mutate(min_r2=min(mean_r_2), max_r2=max(mean_r_2)) %>%
  ungroup()




  

yaxis_set_zero_plot<-summarized_constant_ld %>%
  mutate(recombination_rate=fct_rev(recombination_rate)) %>%
  filter(recombination_rate != "r=1e-07" & recombination_rate != "r=1e-06") %>%
  ggplot(aes(x=distance_breaks, y=mean_r_2, color=factor(selection_coefficient)) )+  
  geom_smooth(method = "gam", formula = y ~ s(x, bs = "cs"), aes(fill=selection_coefficient)) +
  theme_bw() +
  labs(x="Distance (kbp)", y=bquote("Mean"~r^2 ), colour="Selection Coefficient" ) +
  facet_wrap(dominance_coefficient ~ recombination_rate, scales="free_y", nrow=2) +
  scale_colour_manual(values=cbPalette) + 
  scale_fill_manual(values=cbPalette) +
  theme(strip.background =element_rect(fill="white"), text = element_text(size = 22))  +
  scale_x_continuous(breaks = scales::pretty_breaks(n=3)) + 
  guides(fill=FALSE)    + 
  scale_y_continuous(expand = c(0, 0), limits = c(0, NA))


ggsave(filename="../figures/si_figure_2_r2_decay.tiff", plot=yaxis_set_zero_plot, width=20, height=12)



```


## SI Fig 3

```{r}
ldTable<-as_tibble(fread("/Users/jessegarcia/Desktop/SLiM_ParallelRProject/data/decomposeGravel/LDTableOfAllACExceptIntergenic.csv"))


ldTable$Scenario<-case_when(

ldTable$Selection == "../data//SelectionYes" ~ "Full Gravel Model",

ldTable$Selection == "../data/decomposeGravel//DFESelectionGravelOnlyYRI" ~ "Population Size of 7,310 with Expansion to 14,474",

ldTable$Selection == "../data/decomposeGravel//DFESelectionGravelNoMigration" ~ "Gravel Model Without Migration",

ldTable$Selection == "../data/decomposeGravel//DFESelectionConstantPopGravel" ~ "Constant Population Size of 14,474"

)


si_fig_3 <- ldTable %>% 
  filter(AC < 100,Scenario == "Constant Population Size of 14,474") %>% 
  group_by(Scenario, AC, Variation) %>% summarise(meanDist=mean(Distance),  meanSel=mean( ((SelectionCoefficient.x+SelectionCoefficient.y)/2)) ) %>% 
  ggplot(aes(x=AC, y=meanSel, colour=Variation)) + geom_line() + geom_point() + theme_bw(base_size=16) + 
  labs(y="Mean Selection Coefficient", x="Allele Count")   

si_fig_3

ggsave(filename="../figures/si_figure_3_mean_selection.tiff", plot=si_fig_3, width=20, height=12)

```



## SI Fig 5
```{r}
library(data.table)
library(tidyverse)

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

colors=c("0"="#0072B2", "-1e-04"="#F0E442", "-0.001"="#009E73", "-0.01"= "#56B4E9", "-0.1"="#E69F00", "DFE"= "#999999")

selection_levels=c("0", "-1e-04", "-0.001", "-0.01", "-0.1", "DFE")
selection_levels=rev(selection_levels)


## Changing the order of dominance levels for this
recombination_levels=c("r=1e-06", "r=1e-07", "r=1e-08" ,"r=1e-09")
dominance_levels=c("h=0.5", "h=0" )

library(data.table)
library(tidyverse)
library(glue)

## This file can be found on the paper's data dryad dataset.
doubletons_ld<-as_tibble(fread("/Users/jessegarcia/Documents/SLiM_ParallelRProject copy/data2/constant_population_constant_selection_constant_dfe_doubletons_all_replicates_ld.csv", data.table = FALSE))

set.seed(1)
subsampled_doubletons<-doubletons_ld %>%
  filter(selection_coefficient != -0.1) %>%
  group_by(recombination_rate, selection_coefficient, dominance_coefficient, seed) %>%
  nest() %>%
  ungroup() %>%
  group_by(recombination_rate, selection_coefficient, dominance_coefficient) %>%
  sample_n(150) %>%
  mutate(group=rep(c(1,2,3), each=50)) %>%
  unnest(data)


subsampled_doubletons_summary<-subsampled_doubletons %>%
  group_by(recombination_rate, selection_coefficient, dominance_coefficient,group) %>%
  summarise(mean_r_2=mean(r_2)) %>% 
  ungroup() %>%
  mutate(selection_coefficient = replace(selection_coefficient, selection_coefficient == -999, "DFE")) %>%
  mutate(selection_coefficient = factor(selection_coefficient, levels=selection_levels) ) %>%
  mutate(dominance_coefficient = glue("h={dominance_coefficient}")) %>%
  mutate(recombination_rate=glue("r={recombination_rate}")) %>%
  mutate(dominance_coefficient=factor(dominance_coefficient,levels= dominance_levels)) %>%
  mutate(recombination_rate=factor(recombination_rate, levels=recombination_levels)) 


subsampled_doubletons_summary_error_bars<-subsampled_doubletons_summary %>%
  group_by(recombination_rate,selection_coefficient,dominance_coefficient) %>%
  summarise(max_r_2=max(mean_r_2), min_r_2=min(mean_r_2), middle_r_2= sort(mean_r_2, decreasing = FALSE)[2] , mean_of_mean_r2=mean(mean_r_2)) %>% 
  ungroup()





si_fig_5 <- subsampled_doubletons_summary_error_bars %>%
  mutate(recombination_rate=fct_rev(recombination_rate)) %>%
  ggplot(aes(x=recombination_rate, y=mean_of_mean_r2, colour=selection_coefficient)) +
  geom_errorbar(aes(x=recombination_rate,ymin = min_r_2, ymax=max_r_2) , size=2 , width=.1,alpha=0.5)  +
  facet_wrap(~dominance_coefficient, nrow = 2) +
  geom_line(aes(group=selection_coefficient)) +
  scale_colour_manual(values=colors) + 
  guides(fill=FALSE)  +
  theme_bw() +
    theme(strip.background =element_rect(fill="white"), text = element_text(size = 18)) +
  labs(x="Recombination rate",y=bquote("Mean"~r^2 ), colour="Selection Coefficient") +
  geom_line(aes(group=selection_coefficient), size=2)

si_fig_5
ggsave(filename="../figures/si_figure_5_mean_r2.tiff", plot=si_fig_5, width=20, height=12)

```



## SI Fig 6

```{r}
library(tidyverse)
library(glue)
library(mltools)

midpoints <- function(x, dp=2){
lower <- as.numeric(gsub(",.*","",gsub("\\(|\\[|\\)|\\]","", x)))
upper <- as.numeric(gsub(".*,","",gsub("\\(|\\[|\\)|\\]","", x)))
return(round(lower+(upper-lower)/2, dp))
}

# Analyzing Lower recombination rate doubletons

lowRecombSLD<-read_rds("/Users/jessegarcia/Documents/SLiM_ParallelRProject/data/gravelMigrationAndNoMigrationSimulations_attemptFour_smallerRecombinations_June10_SLD.rds")



NSLD<-lowRecombSLD %>% select(recombinationRate, seed, migration, NSLD) %>% unnest(NSLD) %>% unnest(unbiasedD)
SLD<-lowRecombSLD %>% select(recombinationRate, seed, migration, SLD) %>% unnest(SLD)
SLD$type<-SLD$unbiasedD %>% map(~ typeof(.x)) %>% unlist()
SLD<-SLD %>% 
  filter(type != "NULL") %>%
  unnest(unbiasedD)

lowRecombLD<-bind_rows(SLD %>% mutate(variation="Synonymous"), NSLD %>% mutate(variation = "Nonsynonymous")) %>% 
  mutate(dprime=case_when(
  d > 0 ~ d/(.02*.98),
  d < 0 ~ d/(.02*.02),
  d == 0 ~ 0
  
  
  
)) %>%
  mutate(recombinationRateNumber=recombinationRate) %>%
  mutate(recombinationRate = glue("r={recombinationRate}")) %>%
  mutate(recombinationRate=fct_rev(recombinationRate)) %>%
  mutate(geneticDistance=distance*recombinationRateNumber*(1/.01)) 


si_fig_6 <- lowRecombLD %>% 
  filter(migration=="YesMigration") %>% 
  group_by(recombinationRateNumber ) %>%
  mutate(geneticDistanceBreaks=as.character(bin_data(geneticDistance, bins=5, binType="quantile"))) %>% 
  ungroup() %>%
  group_by(recombinationRate, variation, geneticDistanceBreaks) %>%
  summarise(completeRepulsion=mean( dprime == -1)) %>%
  mutate(geneticDistanceMidpoints=midpoints(geneticDistanceBreaks, dp = 30)) %>%
  ggplot(aes(x=geneticDistanceMidpoints, y=completeRepulsion, colour=variation)) +
  geom_point() +
  facet_grid(~ recombinationRate, scales = "free") +
  geom_line()+ 
  labs(x="Genetic Distance Midpoints", y="Fraction of Doubletons with complete repulsion (D' = -1)", colour="Variation" )  + 
  theme_bw() +
  theme(text=element_text(size=18))



si_fig_6
ggsave(filename="figures/si_figure_6_fraction_repulsion.tiff", plot=si_fig_6, width=20, height=12)

```

## SI Fig 8
```{r}
library(data.table)
library(tidyverse)
library(glue)
library(mltools)

ld<-fread("/Users/jessegarcia/Documents/SLiM_ParallelRProject/data/doubletonsLD_gravel_migration_April19.csv")
ld<-as_tibble(ld) %>% mutate(migration=case_when(
  migration=="NoMigration" ~ "No Migration", 
  migration == "YesMigration" ~ "Migration"
)) 

ld<-ld %>% mutate(dprime=case_when(
  d > 0 ~ d/(.02*.98),
  d < 0 ~ d/(.02*.02)
  
  
  
  
)) %>%
  mutate(recombinationRate = glue("r={recombinationRate}")) %>%
  mutate(recombinationRate=fct_rev(recombinationRate)) %>%
  as_tibble()



si_fig_8 <- ld %>% 
  ggplot(aes(x=distance, y=d, colour=variation, linetype=migration)) +
  geom_smooth() +
  facet_wrap(~ recombinationRate) + 
  xlim(0,100000) +
  labs(x="Distance (bp)", y="D", linetype="Migration", colour="Variation") + 
  scale_color_manual(values=c("Synonymous"="dodgerblue1", "Nonsynonymous"="darkorchid2")) +
   theme(strip.background =element_rect(fill="white"), text = element_text(size=18)) +
  scale_x_continuous(breaks = scales::pretty_breaks(n=3), labels = scales::scientific) +
  theme_bw()
si_fig_8
ggsave(filename="figures/si_figure_8_d_migration_no_migration.tiff", plot=si_fig_8, width=20, height=12)

```





## SI Fig 9

```{r}
library(tidyverse)
library(data.table)
library(glue)


label_translation<-c("NoMigration" = "Model 3 \n(No migration)", "YesMigration" = "Model 2 \n(Migration)")

cbbPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

cbbPalette_labels <- c("African"="#56B4E9", "African (no migration)"="#009E73","East Asian"="#F0E442", "European" = "#0072B2")

infoOfVariants<-read_rds("/Users/jessegarcia/Desktop/SLiM_ParallelRProject/data2/gravel_migration_no_migration_variant_metadata.rds") %>% ungroup() 


si_fig_9<-infoOfVariants %>% 
  mutate(recombinationRate = case_when(
    recombinationRate == 1e-09 ~ "r=1e-09", 
    recombinationRate == 1e-08 ~ "r=1e-08"
  )) %>%
  mutate(recombinationRate=factor(recombinationRate, levels=c("r=1e-09", "r=1e-08"))) %>%
  filter(ac==2) %>%
  group_by(migration, seed,extra_po, recombinationRate ) %>%
  summarise(mean_sel=mean(selection_coefficient)) %>% 
  ungroup() %>%
  mutate(extra_po = case_when(
    extra_po == "No Migration" ~ "African (no migration)", 
    TRUE ~ extra_po
  )) %>%
  mutate(pretty_migration = factor(label_translation[migration]) )%>%
  mutate(pretty_migration = fct_rev(pretty_migration) )%>%
  ggplot(aes(x=pretty_migration, y=mean_sel, fill=extra_po)) +
           geom_boxplot(position = position_dodge(preserve = "single")) + 
  labs(x="", y="Mean selection coefficient", fill="Population origin") +
  
##  scale_x_discrete(breaks=label_translation) +
  facet_wrap(~recombinationRate) +
  scale_fill_manual(values=cbbPalette_labels) +
    theme_bw() +
  theme(text=element_text(size=22))


ggsave(filename="figures/si_figure_9_selection_migration_no_migration.tiff", plot=si_fig_9, width=20, height=12)



```

